prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs

```diff
@@ -0,0 +1,191 @@
+    for pred in predicates {
+        match pred {
+            Predicate::Struct(StructPredicate { body: Some(e), .. }) => {
+                ExprWalker::walk(&mut collector, e)
+            }
+            Predicate::Struct(_) => { /* ignore */ }
+            Predicate::Enum(p) => {
+                ExprWalker::walk(&mut collector, &p.discriminant);
+                ExprWalker::walk(&mut collector, &p.discriminant_bounds);
+
+                for (e, _, sp) in &p.variants {
+                    ExprWalker::walk(&mut collector, e);
+                    collector.used_predicates.insert(sp.name.to_string());
+                    sp.body
+                        .iter()
+                        .for_each(|e| ExprWalker::walk(&mut collector, e))
+                }
+            }
+            Predicate::Bodyless(_, _) => { /* ignore */ }
+        }
+    }
+    collector.used_predicates
+}
+
+fn remove_body_of_predicates_if_possible(
+    predicates: &[Predicate],
+    predicates_only_used_in_predicates: &BTreeSet<String>,
+) -> Vec<Predicate> {
+    let mut new_predicates = predicates.to_vec();
+
+    new_predicates.iter_mut().for_each(|predicate| {
+        let predicates_used_in_this_predicate =
+            get_used_predicates_in_predicates(&[predicate.clone()]);
+        if predicates_used_in_this_predicate
+            .intersection(&predicates_only_used_in_predicates)
+            .next()
+            .is_some()
+        {
+            if let Predicate::Struct(sp) = predicate {
+                sp.body = None;
+            }
+        }
+    });
+    new_predicates
+}
+
+pub fn delete_unused_predicates(
+    methods: &[CfgMethod],
+    functions: &[Function],
+    predicates: &[Predicate],
+) -> Vec<Predicate> {
+    let mut has_changed = true;
+    let mut new_predicates: Vec<Predicate> = predicates.to_vec();
+
+    let used_preds = get_used_predicates(methods, functions);
+
+    debug!(
+        "The used predicates in functions and methods are {:?}",
+        &used_preds
+    );
+
+    while has_changed {
+        has_changed = false;
+
+        let predicates_used_in_predicates = get_used_predicates_in_predicates(&new_predicates);
+        debug!(
+            "The used predicates in predicates are {:?}",
+            &predicates_used_in_predicates
+        );
+        new_predicates = new_predicates
+            .into_iter()
+            .filter(|p| {
+                let name = p.name();
+                let is_used_in_predicate = predicates_used_in_predicates.contains(name);
+                let is_used_in_func_or_method = used_preds.contains(name);
+                let is_used = is_used_in_predicate || is_used_in_func_or_method;
+                if !is_used {
+                    debug!("The predicate {} was never used and thus removed", name);
+                    has_changed = true;
+                }
+
+                is_used
+            })
+            .collect();
+    }
+
+    // FIXME: This acctually removes bodies that are needed
```

Comment:
I would prefer to avoid the fixpoint if possible because it could void the gains of the optimization.

Have you checked where the predicates with removed bodies are used? Are they used only in `fold`, `unfold`, and `unfolding`? If yes, we could collect the predicates whose folding state is changed in a special list and remove only the bodies of predicates that are not mentioned in that list.

Reply:
In other words: could the optimization implemented in `prusti-common/src/vir/optimizations/predicates/remove_unnecessary_bodies.rs` be merged into this file?

Reply:
I rearranged the optimisations such that remove_unnecessary_bodies is applied first. Even in that case it appears that around 5 of the tests in the test suite require more than 1 iteration to remove all unused predicates. Should I remove the fixpoint iteration since in most cases most unused predicates are removed that way or do we aim to really eliminate all unused predicates?

Reply:
When you remove an unused predicates `P` you know precisely which other predicates might become unused: they are the predicates that were used by `P`. So, you can keep a much smaller set of predicates to be checked in the next iteration.